diff --git a/kernel/heartbeat_stubs.c b/kernel/heartbeat_stubs.c
index e17d9ee..2c477e2 100644
--- a/kernel/heartbeat_stubs.c
+++ b/kernel/heartbeat_stubs.c
@@ -30,10 +30,18 @@ static void *heartbeat_thread(void *interval_us) {
 
   while (true) {
     struct timespec remain = {0};
+
+#ifdef __APPLE__
+    int err = nanosleep(&interval, &remain);
+    while (err == EINTR) {
+      err = nanosleep(&remain, &remain);
+    }
+#else
     int err = clock_nanosleep(CLOCK_MONOTONIC, 0, &interval, &remain);
     while (err == EINTR) {
       err = clock_nanosleep(CLOCK_MONOTONIC, 0, &remain, &remain);
     }
+#endif
 
     if (err) {
       caml_fatal_error("Heartbeat thread failed to sleep: %d\n", err);
