--- a/compiler/lib-wasm/wat_output.ml
+++ b/compiler/lib-wasm/wat_output.ml
@@ -536,6 +536,11 @@
         ]
     | Event Parse_info.{ src = None | Some ""; _ } -> [ Comment "@" ]
     | Event Parse_info.{ src = Some src; col; line; _ } ->
+        let src =
+            match Build_path_prefix_map.get_build_path_prefix_map () with
+            | Some map -> Build_path_prefix_map.rewrite map src
+            | None -> src
+        in
         let loc = Format.sprintf "%s:%d:%d" src line col in
         [ Comment ("@ " ^ loc) ]
   and instructions l = List.concat (List.map ~f:instruction l) in
