diff --git a/ocaml/configure b/ocaml/configure
index c9f696847..4a5cf8c2c 100755
--- a/ocaml/configure
+++ b/ocaml/configure
@@ -19863,12 +19863,7 @@ if test "x$ax_pthread_ok" = "xyes"; then
     saved_LIBS="$LIBS"
     CFLAGS="$CFLAGS $PTHREAD_CFLAGS"
     LIBS="$LIBS $PTHREAD_LIBS"
-    ac_fn_c_check_func "$LINENO" "sigwait" "ac_cv_func_sigwait"
-if test "x$ac_cv_func_sigwait" = xyes
-then :
-  printf "%s\n" "#define HAS_SIGWAIT 1" >>confdefs.h
-
-fi
+    printf "%s\n" "#define HAS_SIGWAIT 1" >>confdefs.h

     LIBS="$saved_LIBS"
     CFLAGS="$saved_CFLAGS"
diff --git a/ocaml/configure.ac b/ocaml/configure.ac
index d1224de9b..fdde9a4cd 100644
--- a/ocaml/configure.ac
+++ b/ocaml/configure.ac
@@ -2132,7 +2132,7 @@ AS_CASE([$host],
     saved_LIBS="$LIBS"
     CFLAGS="$CFLAGS $PTHREAD_CFLAGS"
     LIBS="$LIBS $PTHREAD_LIBS"
-    AC_CHECK_FUNC([sigwait], [AC_DEFINE([HAS_SIGWAIT])])
+    AC_DEFINE([HAS_SIGWAIT])
     LIBS="$saved_LIBS"
     CFLAGS="$saved_CFLAGS"],
     [AC_MSG_ERROR(m4_normalize([POSIX threads are required but not supported on
