diff --git a/boot/bootstrap.ml b/boot/bootstrap.ml
--- a/boot/bootstrap.ml
+++ b/boot/bootstrap.ml
@@ -92,7 +92,7 @@ let () =
           issues *)
        (if v < (4, 10, 1) then "-custom" else "-output-complete-exe")
        prog
-       (if v >= (5, 0, 0) then "-I +unix " else "")
+       "-I +unix "
        (List.map modules ~f:(fun m -> m ^ ".ml") |> String.concat ~sep:" "));
   let args = List.tl (Array.to_list Sys.argv) in
   let args =
diff --git a/boot/duneboot.ml b/boot/duneboot.ml
--- a/boot/duneboot.ml
+++ b/boot/duneboot.ml
@@ -616,8 +616,7 @@ end = struct
   let output_complete_obj_arg =
     if ocaml_version < (4, 10) then "-custom" else "-output-complete-exe"
 
-  let unix_library_flags =
-    if ocaml_version >= (5, 0) then [ "-I"; "+unix" ] else []
+  let unix_library_flags = [ "-I"; "+unix" ]
 end
 
 let insert_header fn ~header =
diff --git a/src/ocaml/version.ml b/src/ocaml/version.ml
--- a/src/ocaml/version.ml
+++ b/src/ocaml/version.ml
@@ -42,10 +42,10 @@ let has_sys_opaque_identity version = version >= (4, 3, 0)
 
 let has_vmthreads version = version < (4, 9, 0)
 
-let has_bigarray_library version = version < (5, 0, 0)
+let has_bigarray_library version = false
 
 let supports_alerts version = version >= (4, 8, 0)
 
-let has_sandboxed_otherlibs version = version >= (5, 0, 0)
+let has_sandboxed_otherlibs version = true
 
 let has_META_files version = version >= (5, 0, 0)
