From 4e77efa87e74330d58ec29c174a79b69c80f166f Mon Sep 17 00:00:00 2001
From: Diana Kalinichenko <dkalinichenko@janestreet.com>
Date: Thu, 31 Aug 2023 16:15:46 +0100
Subject: [PATCH] fix dune dependencies

Signed-off-by: Diana Kalinichenko <dkalinichenko@janestreet.com>
---
 ast/dune | 4 ++--
 src/dune | 4 +++-
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/ast/dune b/ast/dune
index 03b233f..d704a50 100644
--- a/ast/dune
+++ b/ast/dune
@@ -5,9 +5,9 @@
 (library
  (name ppxlib_ast)
  (public_name ppxlib.ast)
- (libraries astlib stdlib-shims)
+ (libraries astlib compiler-libs.common ocaml-compiler-libs.common ocaml-compiler-libs.shadow stdlib-shims)
  (flags
-  (:standard -safe-string)
+  (:standard -safe-string -open Ocaml_shadow)
   -w
   -9-27-32)
  (preprocess
diff --git a/src/dune b/src/dune
index a9152b2..977c342 100644
--- a/src/dune
+++ b/src/dune
@@ -3,6 +3,8 @@
  (public_name ppxlib)
  (libraries
   (re_export ppxlib_ast)
+  compiler-libs.common
+  ocaml-compiler-libs.common
   ocaml-compiler-libs.shadow
   astlib
   ppxlib_print_diff
@@ -12,7 +14,7 @@
   stdlib-shims
   sexplib0)
  (flags
-  (:standard -safe-string))
+  (:standard -safe-string -open Ocaml_shadow))
  (ppx.driver
   (main Ppxlib.Driver.standalone)
   (flags
-- 
2.39.3

