--- a/compiler/lib/specialize_js.ml
+++ b/compiler/lib/specialize_js.ml
@@ -24,6 +24,12 @@

 let specialize_instr ~target info i =
   match i, target with
+  | Let (x, Prim (Extern "caml_make_unboxed_int32_vect_bytecode", [y])), _ ->
+      Let (x, Prim (Extern "caml_make_vect", [y; Pc (Int32 0l)]))
+  | Let (x, Prim (Extern "caml_make_unboxed_int64_vect_bytecode", [y])), _ ->
+      Let (x, Prim (Extern "caml_make_vect", [y; Pc (Int64 0L)]))
+  | Let (x, Prim (Extern "caml_make_unboxed_nativeint_vect_bytecode", [y])), _ ->
+      Let (x, Prim (Extern "caml_make_vect", [y; Pc (NativeInt 0l)]))
   | Let (x, Prim (Extern "caml_format_int", [ y; z ])), `JavaScript -> (
       match the_string_of info y with
       | Some "%d" -> (
