--- a/lib/js_of_ocaml/json.ml
+++ b/lib/js_of_ocaml/json.ml
@@ -98,6 +98,10 @@
           let i : int64 = Obj.obj v in
           write_int64 b i
       | id -> failwith (Printf.sprintf "Json.output: unsupported custom value %s " id)
+    else if t = Obj.abstract_tag
+      then
+        (* Presumably a JavaScript value *)
+        Buffer.add_string b (Js.to_string (Unsafe.global##._JSON##stringify v))
     else failwith (Printf.sprintf "Json.output: unsupported tag %d " t)

 let to_json v =
