--- a/compiler/tests-jsoo/gh_1307.ml
+++ b/compiler/tests-jsoo/gh_1307.ml
@@ -15,7 +15,7 @@
   test "a";
   [%expect {|
     input: "a"
-    Stdlib.Parsing.Parse_error
+    Stdlib__Parsing.Parse_error
     failure |}];
   test "aa";
   [%expect {|
@@ -25,7 +25,7 @@
   test "aaa";
   [%expect {|
     input: "aaa"
-    Stdlib.Parsing.Parse_error
+    Stdlib__Parsing.Parse_error
     failure |}];
   let (_ : bool) = Parsing.set_trace old in
   ()
