--- a/compiler/ppx/ppx_optcomp_light.ml
+++ b/compiler/ppx/ppx_optcomp_light.ml
@@ -152,7 +152,7 @@
                     | { pexp_desc = Pexp_tuple l; _ } ->
                         let l =
                           List.map l ~f:(function
-                              | { pexp_desc = Pexp_constant (Pconst_integer (d, None))
+                              | None, { pexp_desc = Pexp_constant (Pconst_integer (d, None))
                                 ; _
                                 } -> int_of_string d
                               | _ -> raise Invalid)
