diff --git a/compiler/lib/ocaml_compiler.ml b/compiler/lib/ocaml_compiler.ml
--- a/compiler/lib/ocaml_compiler.ml
+++ b/compiler/lib/ocaml_compiler.ml
@@ -41,9 +41,9 @@ let rec constant_of_const : _ -> Code.co

 let rec find_loc_in_summary ident' = function
   | Env.Env_empty -> None
-  | Env.Env_value (_summary, ident, description) when Poly.(ident = ident') ->
+  | Env.Env_value (_summary, ident, description, _) when Poly.(ident = ident') ->
       Some description.Types.val_loc
-  | Env.Env_value (summary, _, _)
+  | Env.Env_value (summary, _, _, _)
   | Env.Env_type (summary, _, _)
   | Env.Env_extension (summary, _, _)
   | Env.Env_module (summary, _, _, _)
