--- a/ast/location_error.ml
+++ b/ast/location_error.ml
@@ -40,3 +40,27 @@ let get_location error =
 
 let of_exn = Astlib.Location.Error.of_exn
 let raise error = raise (Astlib.Location.Error error)
+
+let of_extension (extension : Ast.extension) =
+  let open Parsetree in
+  let parse_msg = function
+    | { pstr_desc =
+          Pstr_eval ({ pexp_desc = Pexp_constant (Pconst_string (msg, _, _)); _ }, [])
+      ; _
+      } -> msg
+    | _ -> "ppxlib: failed to extract message in ocaml.error"
+  in
+  let parse_sub_msg = function
+    | { pstr_desc =
+          Pstr_extension (({ txt = "error" | "ocaml.error"; loc }, PStr [ msg ]), [])
+      ; _
+      } -> loc, parse_msg msg
+    | { pstr_loc = loc; _ } -> loc, "ppxlib: failed to parse ocaml.error sub messages"
+  in
+  match extension with
+  | { txt = "error" | "ocaml.error"; loc }, PStr (main :: sub) ->
+    let main = parse_msg main in
+    let sub = List.map parse_sub_msg sub in
+    Some (make ~loc main ~sub)
+  | _ -> None
+;;
