diff --git a/compiler/lib/wasm/wa_binaryen.ml b/compiler/lib/wasm/wa_binaryen.ml
O1 is changed because binaryen `-O2` is too slow. We probably want to undo this change
once separate compilation is supported.
skip-pass=inlining-optimizing is added to `-O3` because it seems to meaningfully improve
startup performance.
--- a/compiler/lib/wasm/wa_binaryen.ml
+++ b/compiler/lib/wasm/wa_binaryen.ml
@@ -91,9 +91,9 @@ let dead_code_elimination
   filter_unused_primitives primitives usage_file

 let optimization_options =
-  [| [ "-O2"; "--skip-pass=inlining-optimizing"; "--traps-never-happen" ]
-   ; [ "-O2"; "--traps-never-happen" ]
-   ; [ "-O3"; "--traps-never-happen" ]
+  [| [ "--simplify-locals-notee-nostructure"; "--vacuum"; "--reorder-locals"] (* --opt=1 *)
+   ; [ "-O2"; "--skip-pass=inlining-optimizing"; "--traps-never-happen" ] (* --opt=2 *)
+   ; [ "-O3"; "--skip-pass=inlining-optimizing"; "--traps-never-happen" ] (* --opt=3 *)
   |]

 let optimize
