--- a/compiler/tests-jsoo/test_marshal_compressed.ml
+++ b/compiler/tests-jsoo/test_marshal_compressed.ml
@@ -16,7 +16,7 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  *)
-
+(*
 let%expect_test _ =
   let data =
     "\132\149\166\189\r\022\206\021\001\147F\137d(\181/\253\000Xm\000\0000\n\
@@ -28,4 +28,4 @@
     else String.make 10000 'c'
   in
   Printf.printf "%s ... (%d)\n" (String.sub s 0 20) (String.length s);
-  [%expect {| cccccccccccccccccccc ... (10000) |}]
+  [%expect {| cccccccccccccccccccc ... (10000) |}] *)
