diff --git a/src/ocaml/utils/language_extension.ml b/src/ocaml/utils/language_extension.ml
index bf673ba1..fc76e8cd 100644
--- a/src/ocaml/utils/language_extension.ml
+++ b/src/ocaml/utils/language_extension.ml
@@ -442,12 +442,4 @@ let _ =
   assert Universe.(is maximal);
   (* It's safe to call this here because we've confirmed that we can. *)
   unconditionally_enable_maximal_without_checks ();
-  (* For now, we override the mode and layouts extensions to be in Beta instead of Alpha.
-     This is to prevent printing unnecessary annotations. *)
-  let lower_mode_extension = function
-  | Pair (Mode, Alpha) -> Pair (Mode, Beta)
-  | Pair (Layouts, Alpha) -> Pair (Layouts, Beta)
-  | _ as pair -> pair
-in
-  extensions := List.map lower_mode_extension !extensions
 ;;
