diff --git a/runtime/unix.c b/runtime/unix.c
index 8bc76f013c..890c2ab793 100644
--- a/runtime/unix.c
+++ b/runtime/unix.c
@@ -541,7 +541,9 @@ void *caml_plat_mem_map(uintnat size, uintnat caml_flags, const char* name)
 {
   uintnat alignment = caml_plat_hugepagesize;
   uintnat reserve_only = caml_flags & CAML_MAP_RESERVE_ONLY;
+#ifdef HAS_HUGE_PAGES
   uintnat no_hugetlb = caml_flags & CAML_MAP_NO_HUGETLB;
+#endif
 #ifdef WITH_ADDRESS_SANITIZER
   return aligned_alloc(alignment, (size + (alignment - 1)) & ~(alignment - 1));
 #else