diff --git a/configure.ac b/configure.ac
index 180d235..d47ab72 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2557,8 +2557,7 @@ AS_CASE([$host],
 # global CFLAGS and LIBS variables. This means that all the subsequent
 # tests can rely on the assumption that pthread is enabled.
 CFLAGS="$CFLAGS $PTHREAD_CFLAGS"
-LIBS="$LIBS $PTHREAD_LIBS"
-AC_CHECK_FUNC([sigwait], [AC_DEFINE([HAS_SIGWAIT])])],
+LIBS="$LIBS $PTHREAD_LIBS"],
 [AC_MSG_ERROR(m4_normalize([POSIX threads are required but not supported on
 this platform]))])])

diff --git a/otherlibs/systhreads/st_posix.h b/otherlibs/systhreads/st_posix.h
index 180d235..d47ab72 100644
--- a/otherlibs/systhreads/st_posix.h
+++ b/otherlibs/systhreads/st_posix.h
@@ -81,7 +81,6 @@ value caml_thread_sigmask(value cmd, value sigs)
 
 value caml_wait_signal(value sigs)
 {
-#ifdef HAS_SIGWAIT
 sigset_t set;
 int retcode, signo;
 
@@ -91,8 +90,4 @@ int retcode, signo;
 caml_leave_blocking_section();
 sync_check_error(retcode, "Thread.wait_signal");
 return Val_int(caml_rev_convert_signal_number(signo));
-#else
-caml_invalid_argument("Thread.wait_signal not implemented");
-return Val_int(0); /* not reached */
-#endif
 }