diff --git a/configure.ac b/configure.ac
index 202e26c2a1..cb3326672b 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2557,8 +2557,7 @@ AS_CASE([$host],
     # global CFLAGS and LIBS variables. This means that all the subsequent
     # tests can rely on the assumption that pthread is enabled.
     CFLAGS="$CFLAGS $PTHREAD_CFLAGS"
-    LIBS="$LIBS $PTHREAD_LIBS"
-    AC_CHECK_FUNC([sigwait], [AC_DEFINE([HAS_SIGWAIT])])],
+    LIBS="$LIBS $PTHREAD_LIBS"],
     [AC_MSG_ERROR(m4_normalize([POSIX threads are required but not supported on
       this platform]))])])
 
diff --git a/otherlibs/systhreads/st_posix.h b/otherlibs/systhreads/st_posix.h
index dccf2935e1..7e2015cec9 100644
--- a/otherlibs/systhreads/st_posix.h
+++ b/otherlibs/systhreads/st_posix.h
@@ -81,7 +81,6 @@ value caml_thread_sigmask(value cmd, value sigs)
 
 value caml_wait_signal(value sigs)
 {
-#ifdef HAS_SIGWAIT
   sigset_t set;
   int retcode, signo;
 
@@ -91,8 +90,4 @@ value caml_wait_signal(value sigs)
   caml_leave_blocking_section();
   sync_check_error(retcode, "Thread.wait_signal");
   return Val_int(caml_rev_convert_signal_number(signo));
-#else
-  caml_invalid_argument("Thread.wait_signal not implemented");
-  return Val_int(0);            /* not reached */
-#endif
 }
diff --git a/otherlibs/systhreads4/st_posix.h b/otherlibs/systhreads4/st_posix.h
index a43c78fbc9..0c6e4daabf 100644
--- a/otherlibs/systhreads4/st_posix.h
+++ b/otherlibs/systhreads4/st_posix.h
@@ -541,7 +541,6 @@ value caml_thread_sigmask(value cmd, value sigs) /* ML */
 
 value caml_wait_signal(value sigs) /* ML */
 {
-#ifdef HAS_SIGWAIT
   sigset_t set;
   int retcode, signo;
 
@@ -551,8 +550,4 @@ value caml_wait_signal(value sigs) /* ML */
   caml_leave_blocking_section();
   st_check_error(retcode, "Thread.wait_signal");
   return Val_int(caml_rev_convert_signal_number(signo));
-#else
-  caml_invalid_argument("Thread.wait_signal not implemented");
-  return Val_int(0);            /* not reached */
-#endif
 }
diff --git a/runtime/caml/s.h.in b/runtime/caml/s.h.in
index 76e971db66..4d0be1ba3a 100644
--- a/runtime/caml/s.h.in
+++ b/runtime/caml/s.h.in
@@ -291,8 +291,6 @@
 
 #undef HAS_POSIX_SPAWN
 
-#undef HAS_SIGWAIT
-
 #undef HAS_HUGE_PAGES
 
 #undef HUGE_PAGE_SIZE
diff --git a/runtime4/caml/s.h.in b/runtime4/caml/s.h.in
index 16ea4b8908..d232be8a09 100644
--- a/runtime4/caml/s.h.in
+++ b/runtime4/caml/s.h.in
@@ -282,8 +282,6 @@
 
 #undef HAS_STACK_OVERFLOW_DETECTION
 
-#undef HAS_SIGWAIT
-
 #undef HAS_HUGE_PAGES
 
 #undef HUGE_PAGE_SIZE
