diff --git a/src/float.ml b/src/float.ml
index 3329161..c5b0bec 100644
--- a/src/float.ml
+++ b/src/float.ml
@@ -4,7 +4,7 @@ external min
   -> (float[@unboxed])
   @@ portable
   = "caml_sse2_float64_min_bytecode" "caml_sse2_float64_min"
-[@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+[@@noalloc]
 
 external max
   :  (float[@unboxed])
@@ -12,7 +12,7 @@ external max
   -> (float[@unboxed])
   @@ portable
   = "caml_sse2_float64_max_bytecode" "caml_sse2_float64_max"
-[@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+[@@noalloc]
 
 module Unboxed = struct
   external min
@@ -21,7 +21,7 @@ module Unboxed = struct
     -> (float#[@unboxed])
     @@ portable
     = "caml_sse2_float64_min_bytecode" "caml_sse2_float64_min"
-  [@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+  [@@noalloc]
 
   external max
     :  (float#[@unboxed])
@@ -29,5 +29,5 @@ module Unboxed = struct
     -> (float#[@unboxed])
     @@ portable
     = "caml_sse2_float64_max_bytecode" "caml_sse2_float64_max"
-  [@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+  [@@noalloc]
 end
diff --git a/src/float.mli b/src/float.mli
index 6cea0ad..6c1db55 100644
--- a/src/float.mli
+++ b/src/float.mli
@@ -17,7 +17,7 @@ external min
   -> (float[@unboxed])
   -> (float[@unboxed])
   = "caml_sse2_float64_min_bytecode" "caml_sse2_float64_min"
-[@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+[@@noalloc]
 
 (** Equivalent to [if x > y then x else y].
 
@@ -28,7 +28,7 @@ external max
   -> (float[@unboxed])
   -> (float[@unboxed])
   = "caml_sse2_float64_max_bytecode" "caml_sse2_float64_max"
-[@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+[@@noalloc]
 
 module Unboxed : sig
   external min
@@ -36,12 +36,12 @@ module Unboxed : sig
     -> (float#[@unboxed])
     -> (float#[@unboxed])
     = "caml_sse2_float64_min_bytecode" "caml_sse2_float64_min"
-  [@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+  [@@noalloc]
 
   external max
     :  (float#[@unboxed])
     -> (float#[@unboxed])
     -> (float#[@unboxed])
     = "caml_sse2_float64_max_bytecode" "caml_sse2_float64_max"
-  [@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+  [@@noalloc]
 end
diff --git a/src/int.ml b/src/int.ml
index 424d434..2bd09a4 100644
--- a/src/int.ml
+++ b/src/int.ml
@@ -30,14 +30,14 @@ module Stubs = struct
     -> (int[@untagged])
     @@ portable
     = "caml_int_popcnt" "caml_int_popcnt_tagged_to_untagged"
-  [@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+  [@@noalloc]
 
   external count_set_bits2
     :  int
     -> int
     @@ portable
     = "caml_int_popcnt" "caml_int_popcnt_untagged_to_untagged"
-  [@@untagged] [@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+  [@@untagged] [@@noalloc]
 
   (** [count_trailing_zeros n] returns the number of least-significant zero bits before
       the least significant set bit in [n]. If [n] is 0, the result is the number of bits
diff --git a/src/int32.ml b/src/int32.ml
index 4e4fd92..17a4840 100644
--- a/src/int32.ml
+++ b/src/int32.ml
@@ -45,7 +45,7 @@ module Stubs = struct
     -> (int32[@unboxed])
     @@ portable
     = "caml_int32_popcnt" "caml_int32_popcnt_unboxed_to_untagged"
-  [@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+  [@@noalloc] 
 
   external shift_left
     :  (int32[@local_opt])
diff --git a/src/int64.ml b/src/int64.ml
index e7f1359..d8b6b17 100644
--- a/src/int64.ml
+++ b/src/int64.ml
@@ -45,8 +45,8 @@ module Stubs = struct
     -> (int64[@unboxed])
     @@ portable
     = "caml_int64_popcnt" "caml_int64_popcnt_unboxed_to_untagged"
-  [@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
-
+  [@@noalloc]
+  
   external shift_left
     :  (int64[@local_opt])
     -> (int64[@local_opt])
diff --git a/src/nativeint.ml b/src/nativeint.ml
index 3e2c135..b5849b0 100644
--- a/src/nativeint.ml
+++ b/src/nativeint.ml
@@ -45,7 +45,7 @@ module Stubs = struct
     -> (nativeint[@unboxed])
     @@ portable
     = "caml_nativeint_popcnt" "caml_nativeint_popcnt_unboxed_to_untagged"
-  [@@noalloc] [@@builtin] [@@no_effects] [@@no_coeffects]
+  [@@noalloc]
 
   external shift_left
     :  (nativeint[@local_opt])
