--- a/toplevel/lib/jsooTop.ml
+++ b/toplevel/lib/jsooTop.ml
@@ -65,6 +65,7 @@
     p := !p + len'';
     len''

+[%%if ocaml_version < (4, 14, 0)]
 let use ffp content =
   let fname, oc =
     Filename.open_temp_file ~mode:[ Open_binary ] "jsoo_toplevel" "fake_stdin"
@@ -78,10 +79,11 @@
   with e ->
     Sys.remove fname;
     raise e
-[@@if ocaml_version < (4, 14, 0)]
+[%%endif]

+[%%if ocaml_version >= (4, 14, 0)]
 let use ffp content = Toploop.use_silently ffp (String content)
-[@@if ocaml_version >= (4, 14, 0)]
+[%%endif]

 let execute printval ?pp_code ?highlight_location pp_answer s =
   let lb = Lexing.from_function (refill_lexbuf s (ref 0) pp_code) in
