--- a/ast/dune
+++ b/ast/dune
@@ -13,7 +13,7 @@
  (preprocess
   (per_module
    ((action
-     (run %{exe:pp/pp.exe} %{ocaml_version} %{input-file}))
+     (run %{exe:pp/pp.exe} 9.99 %{input-file}))
     versions)))
  (lint
   (pps ppxlib_traverse -deriving-keep-w32=impl)))
--- a/astlib/dune
+++ b/astlib/dune
@@ -10,7 +10,7 @@
 (rule
  (targets ast-version)
  (action
-  (run %{ocaml} %{dep:config/gen.ml} %{ocaml_version})))
+  (run %{ocaml} %{dep:config/gen.ml} 9.99)))

 (cinaps
  (files *.ml *.mli)
--- a/ast/dune
+++ b/ast/dune
@@ -5,9 +5,9 @@
 (library
  (name ppxlib_ast)
  (public_name ppxlib.ast)
- (libraries astlib stdlib-shims)
+ (libraries astlib compiler-libs.common ocaml-compiler-libs.common ocaml-compiler-libs.shadow stdlib-shims)
  (flags
-  (:standard -safe-string)
+  (:standard -safe-string -open Ocaml_shadow)
   -w
   -9-27-32)
  (preprocess
--- a/src/dune
+++ b/src/dune
@@ -3,16 +3,19 @@
  (public_name ppxlib)
  (libraries
   (re_export ppxlib_ast)
+  compiler-libs.common
+  ocaml-compiler-libs.common
   ocaml-compiler-libs.shadow
   astlib
   ppxlib_print_diff
   ppx_derivers
+  ppxlib_jane
   ppxlib_traverse_builtins
   stdppx
   stdlib-shims
   sexplib0)
  (flags
-  (:standard -safe-string))
+  (:standard -safe-string -open Ocaml_shadow))
  (ppx.driver
   (main Ppxlib.Driver.standalone)
   (flags
--- a/src/gen/dune
+++ b/src/gen/dune
@@ -2,4 +2,4 @@
  (names gen_ast_pattern gen_ast_builder)
  (flags
   (:standard -safe-string))
- (libraries ppxlib_ast astlib ppxlib_traverse_builtins stdppx stdlib-shims))
+ (libraries ppxlib_ast ppxlib_jane astlib ppxlib_traverse_builtins stdppx stdlib-shims))
--- a/traverse/dune
+++ b/traverse/dune
@@ -4,6 +4,6 @@
  (kind ppx_deriver)
  (flags
   (:standard -safe-string))
- (libraries ppxlib ppxlib_ast ppxlib_traverse_builtins stdppx stdlib-shims)
+ (libraries ppxlib ppxlib_ast ppxlib_jane ppxlib_traverse_builtins stdppx stdlib-shims)
  (preprocess
   (pps ppxlib_metaquot)))
