--- a/ast/versions.ml
+++ b/ast/versions.ml
@@ -570,90 +570,6 @@ end
       n n n;
   )
 *)
-module OCaml_402 = struct
-  module Ast = Astlib.Ast_402
-  include Make_witness(Astlib.Ast_402)
-  let version = 402
-  let string_version = "4.02"
-end
-let ocaml_402 : OCaml_402.types ocaml_version = (module OCaml_402)
-module OCaml_403 = struct
-  module Ast = Astlib.Ast_403
-  include Make_witness(Astlib.Ast_403)
-  let version = 403
-  let string_version = "4.03"
-end
-let ocaml_403 : OCaml_403.types ocaml_version = (module OCaml_403)
-module OCaml_404 = struct
-  module Ast = Astlib.Ast_404
-  include Make_witness(Astlib.Ast_404)
-  let version = 404
-  let string_version = "4.04"
-end
-let ocaml_404 : OCaml_404.types ocaml_version = (module OCaml_404)
-module OCaml_405 = struct
-  module Ast = Astlib.Ast_405
-  include Make_witness(Astlib.Ast_405)
-  let version = 405
-  let string_version = "4.05"
-end
-let ocaml_405 : OCaml_405.types ocaml_version = (module OCaml_405)
-module OCaml_406 = struct
-  module Ast = Astlib.Ast_406
-  include Make_witness(Astlib.Ast_406)
-  let version = 406
-  let string_version = "4.06"
-end
-let ocaml_406 : OCaml_406.types ocaml_version = (module OCaml_406)
-module OCaml_407 = struct
-  module Ast = Astlib.Ast_407
-  include Make_witness(Astlib.Ast_407)
-  let version = 407
-  let string_version = "4.07"
-end
-let ocaml_407 : OCaml_407.types ocaml_version = (module OCaml_407)
-module OCaml_408 = struct
-  module Ast = Astlib.Ast_408
-  include Make_witness(Astlib.Ast_408)
-  let version = 408
-  let string_version = "4.08"
-end
-let ocaml_408 : OCaml_408.types ocaml_version = (module OCaml_408)
-module OCaml_409 = struct
-  module Ast = Astlib.Ast_409
-  include Make_witness(Astlib.Ast_409)
-  let version = 409
-  let string_version = "4.09"
-end
-let ocaml_409 : OCaml_409.types ocaml_version = (module OCaml_409)
-module OCaml_410 = struct
-  module Ast = Astlib.Ast_410
-  include Make_witness(Astlib.Ast_410)
-  let version = 410
-  let string_version = "4.10"
-end
-let ocaml_410 : OCaml_410.types ocaml_version = (module OCaml_410)
-module OCaml_411 = struct
-  module Ast = Astlib.Ast_411
-  include Make_witness(Astlib.Ast_411)
-  let version = 411
-  let string_version = "4.11"
-end
-let ocaml_411 : OCaml_411.types ocaml_version = (module OCaml_411)
-module OCaml_412 = struct
-  module Ast = Astlib.Ast_412
-  include Make_witness(Astlib.Ast_412)
-  let version = 412
-  let string_version = "4.12"
-end
-let ocaml_412 : OCaml_412.types ocaml_version = (module OCaml_412)
-module OCaml_413 = struct
-  module Ast = Astlib.Ast_413
-  include Make_witness(Astlib.Ast_413)
-  let version = 413
-  let string_version = "4.13"
-end
-let ocaml_413 : OCaml_413.types ocaml_version = (module OCaml_413)
 module OCaml_414 = struct
   module Ast = Astlib.Ast_414
   include Make_witness(Astlib.Ast_414)
@@ -668,33 +584,21 @@ module OCaml_500 = struct
   let string_version = "5.0"
 end
 let ocaml_500 : OCaml_500.types ocaml_version = (module OCaml_500)
-module OCaml_501 = struct
-  module Ast = Astlib.Ast_501
-  include Make_witness(Astlib.Ast_501)
-  let version = 501
-  let string_version = "5.1"
+module OCaml_999 = struct
+  module Ast = Astlib.Ast_999
+  include Make_witness(Astlib.Ast_999)
+  let version = 999
+  let string_version = "9.99"
 end
-let ocaml_501 : OCaml_501.types ocaml_version = (module OCaml_501)
+let ocaml_999 : OCaml_999.types ocaml_version = (module OCaml_999)
 (*$*)
 
 let all_versions : (module OCaml_version) list = [
   (*$foreach_version (fun n _ ->
       printf "(module OCaml_%d : OCaml_version);\n" n)*)
-(module OCaml_402 : OCaml_version);
-(module OCaml_403 : OCaml_version);
-(module OCaml_404 : OCaml_version);
-(module OCaml_405 : OCaml_version);
-(module OCaml_406 : OCaml_version);
-(module OCaml_407 : OCaml_version);
-(module OCaml_408 : OCaml_version);
-(module OCaml_409 : OCaml_version);
-(module OCaml_410 : OCaml_version);
-(module OCaml_411 : OCaml_version);
-(module OCaml_412 : OCaml_version);
-(module OCaml_413 : OCaml_version);
 (module OCaml_414 : OCaml_version);
 (module OCaml_500 : OCaml_version);
-(module OCaml_501 : OCaml_version);
+(module OCaml_999 : OCaml_version);
 (*$*)
 ]
 
@@ -703,34 +607,10 @@ let all_versions : (module OCaml_version
     printf "    (Astlib.Migrate_%d_%d)(Astlib.Migrate_%d_%d)\n" a b b a
   )
 *)
-include Register_migration(OCaml_402)(OCaml_403)
-    (Astlib.Migrate_402_403)(Astlib.Migrate_403_402)
-include Register_migration(OCaml_403)(OCaml_404)
-    (Astlib.Migrate_403_404)(Astlib.Migrate_404_403)
-include Register_migration(OCaml_404)(OCaml_405)
-    (Astlib.Migrate_404_405)(Astlib.Migrate_405_404)
-include Register_migration(OCaml_405)(OCaml_406)
-    (Astlib.Migrate_405_406)(Astlib.Migrate_406_405)
-include Register_migration(OCaml_406)(OCaml_407)
-    (Astlib.Migrate_406_407)(Astlib.Migrate_407_406)
-include Register_migration(OCaml_407)(OCaml_408)
-    (Astlib.Migrate_407_408)(Astlib.Migrate_408_407)
-include Register_migration(OCaml_408)(OCaml_409)
-    (Astlib.Migrate_408_409)(Astlib.Migrate_409_408)
-include Register_migration(OCaml_409)(OCaml_410)
-    (Astlib.Migrate_409_410)(Astlib.Migrate_410_409)
-include Register_migration(OCaml_410)(OCaml_411)
-    (Astlib.Migrate_410_411)(Astlib.Migrate_411_410)
-include Register_migration(OCaml_411)(OCaml_412)
-    (Astlib.Migrate_411_412)(Astlib.Migrate_412_411)
-include Register_migration(OCaml_412)(OCaml_413)
-    (Astlib.Migrate_412_413)(Astlib.Migrate_413_412)
-include Register_migration(OCaml_413)(OCaml_414)
-    (Astlib.Migrate_413_414)(Astlib.Migrate_414_413)
 include Register_migration(OCaml_414)(OCaml_500)
     (Astlib.Migrate_414_500)(Astlib.Migrate_500_414)
-include Register_migration(OCaml_500)(OCaml_501)
-    (Astlib.Migrate_500_501)(Astlib.Migrate_501_500)
+include Register_migration(OCaml_500)(OCaml_999)
+    (Astlib.Migrate_500_999)(Astlib.Migrate_999_500)
 (*$*)
 
 module OCaml_current = OCaml_OCAML_VERSION
