--- a/src/gen/import.ml
+++ b/src/gen/import.ml
@@ -38,7 +38,12 @@ let common_prefix l =
 
 let map_keyword = function
   | ( "open" | "private" | "downto" | "to" | "mutable" | "rec" | "nonrec"
-    | "virtual" | "type" | "mod" | "begin" | "end" ) as s ->
+    | "virtual" | "type" | "mod" | "begin" | "end" | "with" | "functor"
+    (* "default" isn't a keyword, but the [Default] constructor used
+       for the jkind annotation conflicts with a very common variable
+       name.
+    *)
+    | "default" ) as s ->
       s ^ "_"
   | s -> s
 
