--- a/ast/versions.mli
+++ b/ast/versions.mli
@@ -142,21 +142,9 @@ end
     printf "module OCaml_%d : OCaml_version with module Ast = Astlib.Ast_%d\n"
       n n
   )*)
-module OCaml_402 : OCaml_version with module Ast = Astlib.Ast_402
-module OCaml_403 : OCaml_version with module Ast = Astlib.Ast_403
-module OCaml_404 : OCaml_version with module Ast = Astlib.Ast_404
-module OCaml_405 : OCaml_version with module Ast = Astlib.Ast_405
-module OCaml_406 : OCaml_version with module Ast = Astlib.Ast_406
-module OCaml_407 : OCaml_version with module Ast = Astlib.Ast_407
-module OCaml_408 : OCaml_version with module Ast = Astlib.Ast_408
-module OCaml_409 : OCaml_version with module Ast = Astlib.Ast_409
-module OCaml_410 : OCaml_version with module Ast = Astlib.Ast_410
-module OCaml_411 : OCaml_version with module Ast = Astlib.Ast_411
-module OCaml_412 : OCaml_version with module Ast = Astlib.Ast_412
-module OCaml_413 : OCaml_version with module Ast = Astlib.Ast_413
 module OCaml_414 : OCaml_version with module Ast = Astlib.Ast_414
 module OCaml_500 : OCaml_version with module Ast = Astlib.Ast_500
-module OCaml_501 : OCaml_version with module Ast = Astlib.Ast_501
+module OCaml_999 : OCaml_version with module Ast = Astlib.Ast_999
 (*$*)
 
 (* An alias to the current compiler version *)
