--- a/ast/ast_helper_lite.mli
+++ b/ast/ast_helper_lite.mli
@@ -150,8 +150,16 @@ module Exp : sig
     pattern ->
     expression ->
     expression
+  [@@alert
+    prefer_jane_syntax
+      "Use [Ppxlib_jane.Ast_builder] for constructing functions ASTs that are handed to \
+       the ocaml compiler, e.g. in ppxes"]
 
   val function_ : ?loc:loc -> ?attrs:attrs -> case list -> expression
+  [@@alert
+     prefer_jane_syntax
+       "Use [Ppxlib_jane.Ast_builder] for constructing function ASTs that are handed to \
+        the ocaml compiler, e.g. in ppxes"]
 
   val apply :
     ?loc:loc ->
@@ -162,6 +170,7 @@ module Exp : sig
 
   val match_ : ?loc:loc -> ?attrs:attrs -> expression -> case list -> expression
   val try_ : ?loc:loc -> ?attrs:attrs -> expression -> case list -> expression
+
   val tuple : ?loc:loc -> ?attrs:attrs -> expression list -> expression
 
   val construct :
@@ -264,6 +273,8 @@ module Exp : sig
 
   val extension : ?loc:loc -> ?attrs:attrs -> extension -> expression
   val unreachable : ?loc:loc -> ?attrs:attrs -> unit -> expression
+  val stack : ?loc:loc -> ?attrs:attrs -> expression -> expression
+
   val case : pattern -> ?guard:expression -> expression -> case
   val binding_op : str -> pattern -> expression -> loc -> binding_op
 end
@@ -351,7 +362,7 @@ module Mty : sig
   val attr : module_type -> attribute -> module_type
   val ident : ?loc:loc -> ?attrs:attrs -> lid -> module_type
   val alias : ?loc:loc -> ?attrs:attrs -> lid -> module_type
-  val signature : ?loc:loc -> ?attrs:attrs -> signature -> module_type
+  val signature : ?loc:loc -> ?attrs:attrs -> signature_item list -> module_type
 
   val functor_ :
     ?loc:loc -> ?attrs:attrs -> functor_parameter -> module_type -> module_type
