--- a/src/deriving.mli
+++ b/src/deriving.mli
@@ -108,10 +108,10 @@ val add :
   ?str_type_ext:(structure, type_extension) Generator.t ->
   ?str_exception:(structure, type_exception) Generator.t ->
   ?str_module_type_decl:(structure, module_type_declaration) Generator.t ->
-  ?sig_type_decl:(signature, rec_flag * type_declaration list) Generator.t ->
-  ?sig_type_ext:(signature, type_extension) Generator.t ->
-  ?sig_exception:(signature, type_exception) Generator.t ->
-  ?sig_module_type_decl:(signature, module_type_declaration) Generator.t ->
+  ?sig_type_decl:(signature_item list, rec_flag * type_declaration list) Generator.t ->
+  ?sig_type_ext:(signature_item list, type_extension) Generator.t ->
+  ?sig_exception:(signature_item list, type_exception) Generator.t ->
+  ?sig_module_type_decl:(signature_item list, module_type_declaration) Generator.t ->
   ?extension:(loc:Location.t -> path:string -> core_type -> expression) ->
   string ->
   t
