--- a/src/ast_builder.mli
+++ b/src/ast_builder.mli
@@ -20,6 +20,8 @@ module Default : sig
 
   include module type of Ast_builder_generated.M
 
+  val pexp_apply : loc:location -> expression -> (arg_label * expression) list -> expression
+
   module Latest : sig
     (** This module contains updated versions of node constructors that were
         kept stable when the node changed. For every function in this module,
@@ -43,6 +45,17 @@ module Default : sig
       res:core_type option ->
       unit ->
       constructor_declaration
+
+    val pmty_signature : loc:location -> signature -> module_type
+
+    val signature : loc:location -> ?modalities:modalities -> signature_item list -> signature
+
+    val label_declaration :
+      loc:location -> name:string loc ->
+      mutable_:mutable_flag ->
+      modalities:modality list ->
+      type_:core_type ->
+      label_declaration
   end
 
   val ppat_construct :
@@ -84,6 +97,16 @@ module type S = sig
       res:core_type option ->
       unit ->
       constructor_declaration
+
+    val pmty_signature : signature -> module_type
+    val signature : ?modalities:modalities -> signature_item list -> signature
+
+    val label_declaration :
+      name:string loc ->
+      mutable_:mutable_flag ->
+      modalities:modality list ->
+      type_:core_type ->
+      label_declaration
   end
 
   val ppat_construct : longident loc -> pattern option -> pattern
